### ISSUE TITLE
No Autocomplete featuire 

### ISSUE THREAD

**Issue by a4arpon** [denoland/deno#27093]
Version: Deno latest

There is no autocomplete feature for packages... Please fix it... 

```json
{
  "tasks": {
    "dev": "deno run --watch main.ts"
  },
  "imports": {
    "@std/assert": "jsr:@std/assert@1",
    "mongoose": "npm:mongoose@^8.8.3"
  },
  "fmt": {
    "indentWidth": 2,
    "useTabs": false,
    "semiColons": false
  }
}
```

My MongoDB package is installed but in my editor autocomplete is not working.. I mean no autocomplete suggestion  in Deno please add this feature

**Comment by bartlomieju** [comment]
Deno LSP does provide auto completion for dependencies. Which editor are you using?

**Comment by marvinhagemeister** [comment]
I think the OP is asking for autocompletion in `deno.json` when filling out dependencies.

Deno, no autocompletion:

<img width="723" alt="Screenshot 2024-11-27 at 12 39 06" src="https://github.com/user-attachments/assets/43709c44-c7ae-49ef-8c35-820c82046f47">

Node:

<img width="695" alt="Screenshot 2024-11-27 at 12 39 46" src="https://github.com/user-attachments/assets/597eea46-2d13-4593-b873-b7cbfdbbf059">

...and when you enter the version field that gets autocompleted too:

<img width="673" alt="Screenshot 2024-11-27 at 12 39 53" src="https://github.com/user-attachments/assets/08b0ec61-01d6-4275-be49-6d0ad01dd85d">

**Comment by a4arpon** [comment]
What I will do with the suggestion in Deno JSON or Package JSON file? I need autocompletion or import suggestions like node or bun in the Editor while coding. For functions, Classes and methods.

**Comment by marvinhagemeister** [comment]
Oh thanks for clearing that up. We weren't sure _where_ you were not getting type information as that info was missing in your original post. In that case I'm confused because we do have those things. Can you share a project where you're not getting any completions? Is the Deno extension enabled? You can run the `Deno: Enable` command in vscode's command palette to force enable it.

**Comment by a4arpon** [comment]
This is my Deno Config File:
![image](https://github.com/user-attachments/assets/62eca409-0f89-416a-962c-3bd62c1b6697)


I am installing dependencies 

![image](https://github.com/user-attachments/assets/29ce284b-304e-48f1-876c-dc77bda0e9f7)


My dependencies are already installed ...


![image](https://github.com/user-attachments/assets/71221a2d-4532-4bdb-912e-fdc766a40b3d)
![image](https://github.com/user-attachments/assets/25ebffd1-9300-4276-9b9a-e10ffb1b3746)


See no suggestions from where to import... 


But if I import anything from this already I am getting the suggestion after importing one thing... 
![image](https://github.com/user-attachments/assets/064e99c9-fdd7-421b-8285-f1298299267a)


But the problem is it is adding another separate line for the import :
![image](https://github.com/user-attachments/assets/277c0852-8f3c-4864-9ee8-e5d8f57bca86)


In Bun or even in Node we just import the Class or function in just one line in the second bracket it adds another separate line. 


These are some serious issues. If I work on a large codebase I might not remember what to import or from where to import. Okay for a second we will consider this. But for new adopters, if I suggest them they will ditch it for these kinds of extra hassle. Cause they are not familiar with this kind of extra hassle. 


Personally, I love Deno more than any other JS projects I can swear But I am frustrated on this topic that  I can't use it for any of my projects because of some small issues and missing features. 

I hope Deno team will focus on these small Developer experience and missing things from Deno..

**Comment by bartlomieju** [comment]
The Deno LSP does provide autocomplete for npm and JSR packages when the workspace is set up correctly, so this is almost certainly an editor setup issue rather than a missing feature.

A few things to check:

- Make sure the Deno extension/integration is **enabled for this workspace** (in VS Code: install the official Deno extension and run "Deno: Enable" / set `"deno.enable": true`). Other editors need their Deno LSP integration configured similarly.
- Install your dependencies so the types are resolvable, e.g. run `deno install` (and if you prefer a local `node_modules`, set `"nodeModulesDir": "auto"` in your `deno.json`).
- `mongoose` ships its own type definitions, so once it's installed and the LSP is enabled, completions should work.

Since there's no specific reproducible defect here and the issue has been waiting on info, I'm closing it. If you've done the above and still get no completions, please open a new issue with your editor, the Deno extension version, and a screenshot, and we'll take a look. Thanks!
